Add a way to visualize memory snapshot traces

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90587
* #90567
* __->__ #90348

This adds a d3-based interactive visualization for exploring the memory
allocation traces that the caching allocator can capture. This visualization
code can also be attached to kineto trace information in the future to also
provide visualization for the memory events captured there, which come with
addition information about the graph.

zdevito committed:
Add a way to visualize memory snapshot traces

This adds a d3-based interactive visualization for exploring the memory
allocation traces that the caching allocator can capture. This visualization
code can also be attached to kineto trace information in the future to also
provide visualization for the memory events captured there, which come with
addition information about the graph.

[ghstack-poisoned]

robieta commented on torch/cuda/_memory_viz.py:
```diff
@@ -301,6 +309,364 @@ def find_segment(addr):
+    return _memory_over_time_template.replace('$PLOT_DATA', plot_data)
+
+
+_memory_over_time_template = r"""
```
For giant templates like this I've found it's better to store them in the native language and then read them in. (e.g. [benchmark cpp template](https://github.com/pytorch/pytorch/blob/master/torch/utils/benchmark/utils/timeit_template.cpp)) That way you get syntax highlighting.

zdevito replied:
I added an inline note here: I agree its better in its own file. Right now I end up download _memory_viz.py to a bunch of places where it would hard to install pytorch (e.g. laptop). I need to figure out a way to make a really small pip installable package for the thing so that it can be multiple files.

robieta commented on torch/cuda/_memory_viz.py:
```diff
@@ -94,6 +100,14 @@ def _seg_info(seg):
+        if abs(num) < 1024.0:
+            return f"{num:3.1f}{unit}B"
+        num /= 1024.0
+    return f"{num:.1f}YiB"
```
Lol

robieta reviewed (approved):
I didn't go over it with a fine tooth comb, but overall looks reasonable.

zdevito committed:
Update on "Add a way to visualize memory snapshot traces"


This adds a d3-based interactive visualization for exploring the memory
allocation traces that the caching allocator can capture. This visualization
code can also be attached to kineto trace information in the future to also
provide visualization for the memory events captured there, which come with
addition information about the graph.

[ghstack-poisoned]

zdevito committed:
Update on "Add a way to visualize memory snapshot traces"


This adds a d3-based interactive visualization for exploring the memory
allocation traces that the caching allocator can capture. This visualization
code can also be attached to kineto trace information in the future to also
provide visualization for the memory events captured there, which come with
addition information about the graph.

[ghstack-poisoned]

zdevito commented:
@pytorchbot merge

eellison commented:
sweet!

zdevito commented:
@pytorchmergebot merge -f "infra failure unrelated"

vadimkantorov commented:
Does this address ` Built-in minimal tool for visualization of used memory (cached memory / used memory) into an SVG/HTML string` in https://github.com/pytorch/pytorch/issues/1529?

zdevito commented:
Yeah, it can be used to give a look at both the used memory and current current block allocations as actions happen in the allocator. Draft of a guide for using this functionality is here: https://zdevito.github.io/2022/12/09/memory-traces.html

vadimkantorov commented:
This is awesome :) Please feel free to comment into that old issue if you consider some of the points completed now
